File: L131-palindrome_partitioning.py
# https://leetcode.com/problems/palindrome-partitioning/
class Solution:
    def isPalindrome(self, s: str) -> bool:
        l, r = 0, len(s) - 1
        while (l <= r):
            if (s[l] != s[r]):
                return False
            l += 1
            r -= 1
        return True
    def bt(self, s: str, dp: List[str], res: List[List[str]]):
        if (s == None or len(s) == 0):
            res.append(dp.copy())
            return
        for i in range(1, len(s) + 1):
            temp = s[0:i]
            if (not self.isPalindrome(temp)):
                continue
            dp.append(temp)
            self.bt(s[i:len(s)], dp, res)
            dp.pop()
    def partition(self, s: str) -> List[List[str]]:
        res = []
        self.bt(s, [], res)
        return res

File: L131-PalindromePartitioning.java
// https://leetcode.com/problems/palindrome-partitioning/
class Solution {
    public Boolean isPalindrome(String s) {
        int l = 0, r = s.length() - 1;
        while (l <= r) {
            if (s.charAt(l) != s.charAt(r)) {
                return false;
            }
            l++;
            r--;
        }
        return true;
    }
    public void backTrack(String s, List<List<String>> res, List<String> dp) {
        if (s == null || s.length() == 0) {
            res.add(new ArrayList<>(dp));
            return;
        }
        for (int i = 1; i <= s.length(); i++) {
            String temp = s.substring(0, i);
            if (!isPalindrome(temp)) {
                continue;
            }
            dp.add(temp);
            backTrack(s.substring(i, s.length()), res, dp);
            dp.remove(dp.size() - 1);
        }
    }
    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<>();
        backTrack(s, res, new ArrayList<>());
        return res;
    }
}
